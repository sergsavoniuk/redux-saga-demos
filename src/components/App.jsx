import React, { useEffect } from 'react';
import { connect } from 'react-redux';
import { push } from 'connected-react-router';

import Routes from 'components/Routes';
import Layout from 'components/Layout';
import AuthForm from 'components/AuthForm';
import Loader from 'components/Loader';
import { login, getIsAuthenticated } from 'redux/auth';

function getUsername() {
  return localStorage.getItem('username');
}

function App({ isAuthenticated, login, push }) {
  useEffect(() => {
    const username = getUsername();
    if (!isAuthenticated) {
      push('/auth');
      if (username != null) {
        login(username);
      }
    }
  }, [isAuthenticated]);

  let Component;
  if (!isAuthenticated) {
    Component = getUsername() != null ? <Loader /> : <AuthForm />;
  } else {
    Component = <Routes />;
  }

  return <Layout>{Component}</Layout>;
}

function mapStateToProps(state) {
  return {
    isAuthenticated: getIsAuthenticated(state),
  };
}

export default connect(
  mapStateToProps,
  { login, push },
)(App);
