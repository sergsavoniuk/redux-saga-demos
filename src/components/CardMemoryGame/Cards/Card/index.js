export { default, CardPropTypes } from './Card';
