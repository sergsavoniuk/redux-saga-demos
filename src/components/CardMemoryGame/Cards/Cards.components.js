import styled, { css } from 'styled-components';

export const Cards = styled.div`
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
  width: 500px;
  height: 500px;
  margin-top: 100px;
  font-family: Arial;

  @media (max-width: 499px) {
    width: 95%;
  }
`;

export const Card = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  width: calc(50% - 2px);
  height: 50%;
  margin: 1px;
  padding: 70px 0;
  background-color: #16a085;
  color: #fff;
  font-size: 3em;
  font-weight: bold;
  cursor: pointer;

  transition-duration: 400ms;
  transition-property: transform;

  &:hover {
    opacity: .8;
  }

  ${props =>
    props.flipped &&
    css`
      flex-direction: column;
      justify-content: flex-start;
      align-items: flex-start;
      width: calc(100% - 2px);
      padding: 0;
      padding: 0 10px;
      background-color: #c0392b;
      font-family: 'Open Sans';
      font-size: 0.85em;
      letter-spacing: 0.8px;
      font-weight: normal;
      transform: rotate(360deg);

      & > p {
        padding: 0;
        margin: 4px 0;
      }

      &:hover {
        opacity: 1;
      }
    `}

  ${props =>
    props.fixed === true &&
    css`
      display: flex;
      flex-direction: column;
      justify-content: space-around;
      padding: 0;
      background-color: #c0392b;

      &:hover {
        opacity: 1;
      }
    `}

  ${props =>
    props.active === false &&
    css`
      background-color: #c0392b;
      transform: rotateY(180deg);

      &:hover {
        opacity: 1;
      }
    `}


`;
