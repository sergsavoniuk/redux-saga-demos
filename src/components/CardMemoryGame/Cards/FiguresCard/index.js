export { default, FiguresCard } from './FiguresCard';
