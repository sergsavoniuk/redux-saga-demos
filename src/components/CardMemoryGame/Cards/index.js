export { default, isCardVisible } from './Cards';
