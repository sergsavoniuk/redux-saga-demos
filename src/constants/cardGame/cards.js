export const CARDS = {
  Figures: 'Figures',
  Mock: 'Mock',
  Instructions: 'Instructions',
  Play: 'Play',
};
