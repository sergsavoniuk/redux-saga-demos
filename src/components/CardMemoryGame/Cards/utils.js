import { CARDS } from 'constants/cardGame/cards';

const { Mock, Figures, Play, Instructions } = CARDS;

function isCardVisible(cardName, flippedCard) {
  if (
    (cardName === Mock && flippedCard === Figures) ||
    (cardName === Play && flippedCard === Instructions)
  ) {
    return false;
  }
  return true;
}

export { isCardVisible };
