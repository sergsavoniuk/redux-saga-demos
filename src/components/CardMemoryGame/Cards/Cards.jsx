import React, { useState } from 'react';

import Card from './Card';
import { Wrapper, Tip, Cards as StyledCards } from './Cards.components';
import { CARDS } from 'constants/cardGame/cards';
import { isCardVisible } from './utils';

function Cards() {
  const [flippedCard, setFlippedCard] = useState(null);

  function handleCardClick(name) {
    if (name === flippedCard) {
      setFlippedCard(null);
    } else {
      setFlippedCard(name);
    }
  }

  return (
    <Wrapper>
      <Tip>Click P to get started.</Tip>
      <StyledCards>
        {Object.keys(CARDS).map(
          name =>
            isCardVisible(CARDS[name], flippedCard) && (
              <Card
                key={name}
                name={CARDS[name]}
                isFlipped={CARDS[name] === flippedCard}
                onCardClick={handleCardClick}
              />
            ),
        )}
      </StyledCards>
    </Wrapper>
  );
}

export default Cards;
