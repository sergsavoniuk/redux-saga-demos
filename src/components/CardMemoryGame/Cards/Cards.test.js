import React from 'react';
import { shallow, mount } from 'enzyme';

import Cards from './Cards';
import CardUtils from './utils';
import { Card } from './Cards.components';
import { CARDS } from 'constants/cardGame/cards';

const { Mock, Figures, Play, Instructions } = CARDS;

jest.mock('./utils', () => {
  return {
    isCardVisible: jest.fn(() => true),
  };
});

beforeEach(() => {
  jest.resetModules();
});

describe('<Cards />', () => {
  test('renders correctly', () => {
    const wrapper = shallow(<Cards />);

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('Card')).toHaveLength(4);
  });

  test('should flip a Figures card after a click and make a Mock card invisible, and after a second click back to the initial state', () => {
    let flippedCard;
    const wrapper = shallow(<Cards />);

    CardUtils.isCardVisible = jest.fn().mockImplementation((cardName, _) => {
      if (cardName === Mock && flippedCard === Figures) {
        return false;
      }
      return true;
    });

    const figuresCard = wrapper
      .find('Card')
      .findWhere(node => node.prop('name') === Figures);
    flippedCard = Figures;
    figuresCard.prop('onCardClick')(Figures);

    expect(wrapper).toMatchSnapshot();

    flippedCard = null;
    figuresCard.prop('onCardClick')(Figures);
    expect(wrapper).toMatchSnapshot();
  });

  test.each`
    cardName | flippedCardName | expected
    ${Mock}  | ${Figures}      | ${false}
    ${Play}  | ${Instructions} | ${false}
    ${Play}  | ${Figures}      | ${true}
  `('test isCardVisible', ({ cardName, flippedCardName, expected }) => {
    const isCardVisible = jest.requireActual('./utils').isCardVisible;
    expect(isCardVisible(cardName, flippedCardName)).toBe(expected);
  });
});

describe('<Card /> styled component', () => {
  test('applies styles when flipped props set to true', () => {
    const wrapper = mount(<Card flipped={true} />);

    expect(wrapper).toHaveStyleRule('flex-direction', 'column');
    expect(wrapper).toHaveStyleRule('justify-content', 'flex-start');
    expect(wrapper).toHaveStyleRule('background-color', '#c0392b');
    expect(wrapper).toHaveStyleRule('font-weight', 'normal');
    expect(wrapper).toHaveStyleRule('opacity', '1', { modifier: ':hover' });
  });

  test('applies styles when fixed props set to true', () => {
    const wrapper = mount(<Card fixed={true} />);

    expect(wrapper).toHaveStyleRule('flex-direction', 'column');
    expect(wrapper).toHaveStyleRule('justify-content', 'space-around');
    expect(wrapper).toHaveStyleRule('background-color', '#c0392b');
    expect(wrapper).toHaveStyleRule('opacity', '1', { modifier: ':hover' });
  });

  test('applies styles when active props set to false', () => {
    const wrapper = mount(<Card active={false} />);

    expect(wrapper).toHaveStyleRule('transform', 'rotateY(180deg)');
    expect(wrapper).toHaveStyleRule('opacity', '1', {
      modifier: ':hover',
    });
  });
});
