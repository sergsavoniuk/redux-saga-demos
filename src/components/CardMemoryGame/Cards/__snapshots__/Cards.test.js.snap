// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Cards /> renders correctly 1`] = `
<styled.div>
  <styled.span>
    Click P to get started.
  </styled.span>
  <styled.div>
    <Card
      isFlipped={false}
      key="Figures"
      name="F"
      onCardClick={[Function]}
    />
    <Card
      isFlipped={false}
      key="Mock"
      name="L"
      onCardClick={[Function]}
    />
    <Card
      isFlipped={false}
      key="Instructions"
      name="I"
      onCardClick={[Function]}
    />
    <Card
      isFlipped={false}
      key="Play"
      name="P"
      onCardClick={[Function]}
    />
  </styled.div>
</styled.div>
`;

exports[`<Cards /> should flip a Figures card after a click and make a Mock card invisible, and after a second click back to the initial state 1`] = `
<styled.div>
  <styled.span>
    Click P to get started.
  </styled.span>
  <styled.div>
    <Card
      isFlipped={true}
      key="Figures"
      name="F"
      onCardClick={[Function]}
    />
    <Card
      isFlipped={false}
      key="Instructions"
      name="I"
      onCardClick={[Function]}
    />
    <Card
      isFlipped={false}
      key="Play"
      name="P"
      onCardClick={[Function]}
    />
  </styled.div>
</styled.div>
`;

exports[`<Cards /> should flip a Figures card after a click and make a Mock card invisible, and after a second click back to the initial state 2`] = `
<styled.div>
  <styled.span>
    Click P to get started.
  </styled.span>
  <styled.div>
    <Card
      isFlipped={true}
      key="Figures"
      name="F"
      onCardClick={[Function]}
    />
    <Card
      isFlipped={false}
      key="Mock"
      name="L"
      onCardClick={[Function]}
    />
    <Card
      isFlipped={false}
      key="Instructions"
      name="I"
      onCardClick={[Function]}
    />
    <Card
      isFlipped={false}
      key="Play"
      name="P"
      onCardClick={[Function]}
    />
  </styled.div>
</styled.div>
`;
